### DISCUSSION TITLE
Gencode from main patternfly repository?

### DISCUSSION THREAD

**Discussion by deathbreakfast** [patternfly-yew/patternfly-yew#11] [Ideas]
I've been playing around with GPT-3 and was seeing how well it would be able to convert React to yew. I think there is an opportunity to increase developer velocity in this repo. I wanted to start the discussion to get the sentiment towards this direction before diving too deep. I should have time to create a prototype next Month.

I tried testing with parts of a component in the official Patternfly repo take for example the following props directly from the [Checkbox](https://github.com/patternfly/patternfly-react/blob/main/packages/react-core/src/components/Checkbox/Checkbox.tsx) component.

```
export interface CheckboxProps
  extends Omit<React.HTMLProps<HTMLInputElement>, 'type' | 'onChange' | 'disabled' | 'label'>,
    OUIAProps {
  /** Additional classes added to the Checkbox. */
  className?: string;
  /** Flag to show if the Checkbox selection is valid or invalid. */
  isValid?: boolean;
  /** Flag to show if the Checkbox is disabled. */
  isDisabled?: boolean;
  /** Flag to show if the Checkbox is checked. If null, the checkbox will be indeterminate (partially checked). */
  isChecked?: boolean | null;
  checked?: boolean;
  /** A callback for when the Checkbox selection changes. */
  onChange?: (checked: boolean, event: React.FormEvent<HTMLInputElement>) => void;
  /** Label text of the checkbox. */
  label?: React.ReactNode;
  /** Id of the checkbox. */
  id: string;
  /** Aria-label of the checkbox. */
  'aria-label'?: string;
  /** Description text of the checkbox. */
  description?: React.ReactNode;
  body?: React.ReactNode;
}
```

It did pretty well here.

```
#[derive(Clone, PartialEq, Properties)]
pub struct CheckboxProps {
  /// Additional classes added to the Checkbox.
  #[props_of_default]
  pub class_name: Option<String>,
  /// Flag to show if the Checkbox selection is valid or invalid.
  #[props_of_default]
  pub is_valid: Option<bool>,
  /// Flag to show if the Checkbox is disabled.
  #[props_of_default]
  pub is_disabled: Option<bool>,
  /// Flag to show if the Checkbox is checked. If null, the checkbox will be indeterminate (partially checked).
  #[props_of_default]
  pub is_checked: Option<bool>,
  /// A callback for when the Checkbox selection changes.
  #[props_of_default]
  pub on_change: Option<Callback<(bool, ChangeData)>>,
  /// Label text of the checkbox.
  #[props_of_default]
  pub label: Option<Children>,
  /// Id of the checkbox.
  pub id: String,
  /// Aria-label of the checkbox.
  #[props_of_default]
  pub aria_label: Option<String>,
  /// Description text of the checkbox.
  #[props_of_default]
  pub description: Option<Children>,
  /// Description text of the checkbox.
  #[props_of_default]
  pub body: Option<Children>,
}

```

This was without any fine tuning all I did was provide a single example followed by a question. 

The question I asked was: `"Convert the TypeScript Props to Yew Props written in the rust programming language. Any react component types such as React.ReactNode or ReactNode can be typed as HTML in rust. TypeScript nullable types are typed as Optional in Rust. And visibility in Rust props are all public.  Variable names should be converted from CamelCase to snake_case"`

And one of the provided examples.

```
React Props:

export interface CardProps extends React.HTMLProps<HTMLElement>, OUIAProps {
  /** Content rendered inside the Card */
  children?: React.ReactNode;
  /** ID of the Card. Also passed back in the CardHeader onExpand callback. */
  id?: string;
  /** Additional classes added to the Card */
  className?: string;
  /** Sets the base component to render. defaults to article */
  component?: keyof JSX.IntrinsicElements;
  /** @deprecated to make a card hoverable, use isSelectable or isSelectableRaised. */
  isHoverable?: boolean;
  /** Modifies the card to include compact styling. Should not be used with isLarge. */
  isCompact?: boolean;
  /** Modifies the card to include selectable styling */
  isSelectable?: boolean;
  /** @beta Specifies the card is selectable, and applies the new raised styling on hover and select */
  isSelectableRaised?: boolean;
  /** Modifies the card to include selected styling */
  isSelected?: boolean;
  /** @beta Modifies a raised selectable card to have disabled styling */
  isDisabledRaised?: boolean;
  /** Modifies the card to include flat styling */
  isFlat?: boolean;
  /** Modifies the card to include rounded styling */
  isRounded?: boolean;
  /** Modifies the card to be large. Should not be used with isCompact. */
  isLarge?: boolean;
  /** Cause component to consume the available height of its container */
  isFullHeight?: boolean;
  /** Modifies the card to include plain styling; this removes border and background */
  isPlain?: boolean;
  /** Flag indicating if a card is expanded. Modifies the card to be expandable. */
  isExpanded?: boolean;
}
```

Yew Props:

```
#[derive(Clone, PartialEq, Properties)]
pub struct CardProps {
  /// Content rendered inside the Card
  pub children: Children,
  /// ID of the Card. Also passed back in the CardHeader onExpand callback.
  #[props_of_default]
  pub id: Option<String>
  /// Additional classes added to the Card
  #[props_of_default]
  pub class_name: Option<String>,
  /// Sets the base component to render. defaults to article
  #[props_of_default]
  pub component: Option<HTML>,
  /// @deprecated to make a card hoverable, use isSelectable or isSelectableRaised.
  #[props_of_default]
  pub is_hoverable: Option<bool>,
  /// Modifies the card to include compact styling. Should not be used with isLarge.
  #[props_of_default]
  pub is_compact: Option<bool>,
  /// Modifies the card to include selectable styling
  #[props_of_default]
  pub is_selectable: Option<bool>,
  /// @beta Specifies the card is selectable, and applies the new raised styling on hover and select
  #[props_of_default]
  pub is_selectable_raised: Option<bool>,
  /// Modifies the card to include selected styling
  #[props_of_default]
  pub is_selected: Option<bool>,
  /// @beta Modifies a raised selectable card to have disabled styling
  #[props_of_default]
  pub is_disabled_raised: Option<bool>,
  /// Modifies the card to include flat styling
  #[props_of_default]
  pub is_flat: Option<bool>,
  /// Modifies the card to include rounded styling
  #[props_of_default]
  pub is_rounded: Option<bool>,
  /// Modifies the card to be large. Should not be used with isCompact.
  #[props_of_default]
  pub is_large: Option<bool>,
  /// Cause component to consume the available height of its container
  #[props_of_default]
  pub is_full_height: Option<bool>,
  /// Modifies the card to include plain styling; this removes border and background
  #[props_of_default]
  pub is_plain: Option<bool>,
  /// Flag indicating if a card is expanded. Modifies the card to be expandable.
  #[props_of_default]
  pub is_expanded: Option<bool>,
}
```

**Comment by deathbreakfast** [comment]
Oops copied wrong example.

Edit: Updated original post with Codex example I used. Using code-davinci-002  instead of text-davinci-002 model.

The thought here is a script can be created, even if it does half of the work correct, there is significant time savings.

**Comment by ctron** [comment]
Amazing. :grinning: 

I like the idea. I guess it depends on how much effort it is. And how people could leverage/maintain that.

Just to be clear, I am not pushing back! I just have so many questions, so discussing this first definitely is the right approach! I will try to structure this a bit.

1. To my understanding, this "only" converts the properties. I would choose different naming here and there, but I guess that is a matter of style and shouldn't be hard to change (Like `selected` instead of `is_selected`). But what about the actual implementation? All those properties need to be applied to the virtual DOM model somehow. And that needs to fit into the actual logic of components. Could GPT-3 handle that too?

1. In some cases I explicitly choose a different path, which felt more Rust native (like using enums for conflicting settings). I know this is a double edged sword, as it deviates from the well documented React implementation, but might be beneficial to prevent the developer from mis-using the code. Could something like this be handled too?

1. What would the workflow look like? Just an "initial" conversion and then manual tracking of changes? Or repeating the generating with every release? What would a developer on this project require to re-generate the code by himself?

I think it is definitely worth to pursue this further. Even if it would just be for exploring the idea, and learning about the practicability of en effort like this. Having lived through a disaster called "model driven development", I am might a burnt child, making me a bit more skeptic than necessary :grin:

**Comment by deathbreakfast** [comment]
I'll break this down into two parts, using the model itself and any kind of product or tool that can leverage it.

#### The Model

To make sure I set the right expectations here, it sucks at Rust. The GitHub training data cuts off in October 2019. It likely has zero context on what Yew is let alone how to use it. So do I expect it to convert an entire file out of the box? No.

Let me take a step back and I'll *briefly* go over how I was using GPT-3 and the type of experimentation that will need to happen to solve this problem.

I was using the [code completion api](https://beta.openai.com/docs/guides/code), through their JS playground. You're limited to up to 4000 words. The performance of the model is dependent on the quality of the prompt or question you ask. It's recommended to provide clear examples or test cases. Here is the full prompt I used.

```
prompt:
Convert the TypeScript Props to Yew Props written in the rust programming language. Any react component types such as React.ReactNode or ReactNode can be typed as HTML in rust. TypeScript nullable types are typed as Optional in Rust. And visibility in Rust props are all public.  Variable names should be converted from CamelCase to snake_case.

TypeScript:

export interface CardProps extends React.HTMLProps<HTMLElement>, OUIAProps {
  /** Content rendered inside the Card */
  children?: React.ReactNode;
  /** ID of the Card. Also passed back in the CardHeader onExpand callback. */
  id?: string;
  /** Additional classes added to the Card */
  className?: string;
  /** Sets the base component to render. defaults to article */
  component?: keyof JSX.IntrinsicElements;
  /** @deprecated to make a card hoverable, use isSelectable or isSelectableRaised. */
  isHoverable?: boolean;
  /** Modifies the card to include compact styling. Should not be used with isLarge. */
  isCompact?: boolean;
  /** Modifies the card to include selectable styling */
  isSelectable?: boolean;
  /** @beta Specifies the card is selectable, and applies the new raised styling on hover and select */
  isSelectableRaised?: boolean;
  /** Modifies the card to include selected styling */
  isSelected?: boolean;
  /** @beta Modifies a raised selectable card to have disabled styling */
  isDisabledRaised?: boolean;
  /** Modifies the card to include flat styling */
  isFlat?: boolean;
  /** Modifies the card to include rounded styling */
  isRounded?: boolean;
  /** Modifies the card to be large. Should not be used with isCompact. */
  isLarge?: boolean;
  /** Cause component to consume the available height of its container */
  isFullHeight?: boolean;
  /** Modifies the card to include plain styling; this removes border and background */
  isPlain?: boolean;
  /** Flag indicating if a card is expanded. Modifies the card to be expandable. */
  isExpanded?: boolean;
}

Rust:

#[derive(Clone, PartialEq, Properties)]
pub struct CardProps {
  /// Content rendered inside the Card
  pub children: Children,
  /// ID of the Card. Also passed back in the CardHeader onExpand callback.
  #[props_of_default]
  pub id: Option<String>
  /// Additional classes added to the Card
  #[props_of_default]
  pub class_name: Option<String>,
  /// Sets the base component to render. defaults to article
  #[props_of_default]
  pub component: Option<HTML>,
  /// @deprecated to make a card hoverable, use isSelectable or isSelectableRaised.
  #[props_of_default]
  pub is_hoverable: Option<bool>,
  /// Modifies the card to include compact styling. Should not be used with isLarge.
  #[props_of_default]
  pub is_compact: Option<bool>,
  /// Modifies the card to include selectable styling
  #[props_of_default]
  pub is_selectable: Option<bool>,
  /// @beta Specifies the card is selectable, and applies the new raised styling on hover and select
  #[props_of_default]
  pub is_selectable_raised: Option<bool>,
  /// Modifies the card to include selected styling
  #[props_of_default]
  pub is_selected: Option<bool>,
  /// @beta Modifies a raised selectable card to have disabled styling
  #[props_of_default]
  pub is_disabled_raised: Option<bool>,
  /// Modifies the card to include flat styling
  #[props_of_default]
  pub is_flat: Option<bool>,
  /// Modifies the card to include rounded styling
  #[props_of_default]
  pub is_rounded: Option<bool>,
  /// Modifies the card to be large. Should not be used with isCompact.
  #[props_of_default]
  pub is_large: Option<bool>,
  /// Cause component to consume the available height of its container
  #[props_of_default]
  pub is_full_height: Option<bool>,
  /// Modifies the card to include plain styling; this removes border and background
  #[props_of_default]
  pub is_plain: Option<bool>,
  /// Flag indicating if a card is expanded. Modifies the card to be expandable.
  #[props_of_default]
  pub is_expanded: Option<bool>,
}

TypeScript:

interface CardContextProps {
  cardId: string;
  /** Modifies the card to display a larger card body. */
  isExpanded: boolean;
}

Rust:

#[derive(Clone, PartialEq, Properties)]
struct CardContextProps {
  card_id: String,
  /// Modifies the card to display a larger card body.
  is_expanded: bool,
}

TypeScript:

export interface Props {
  /** index of the outer loop */
  idx: number;
}

Rust:

#[derive(Clone, PartialEq, Properties)]
pub struct Props {
  /// index of the outer loop
  pub idx: i64,
}

TypeScript:

export interface CheckboxProps
  extends Omit<React.HTMLProps<HTMLInputElement>, 'type' | 'onChange' | 'disabled' | 'label'>,
    OUIAProps {
  /** Additional classes added to the Checkbox. */
  className?: string;
  /** Flag to show if the Checkbox selection is valid or invalid. */
  isValid?: boolean;
  /** Flag to show if the Checkbox is disabled. */
  isDisabled?: boolean;
  /** Flag to show if the Checkbox is checked. If null, the checkbox will be indeterminate (partially checked). */
  isChecked?: boolean | null;
  checked?: boolean;
  /** A callback for when the Checkbox selection changes. */
  onChange?: (checked: boolean, event: React.FormEvent<HTMLInputElement>) => void;
  /** Label text of the checkbox. */
  label?: React.ReactNode;
  /** Id of the checkbox. */
  id: string;
  /** Aria-label of the checkbox. */
  'aria-label'?: string;
  /** Description text of the checkbox. */
  description?: React.ReactNode;
  body?: React.ReactNode;
}

Rust:

```  

As you can see I added a lot of examples for it to use, if this was in JS or python, it wouldn't need as much context. These examples end up taking the majority of the API word limit. That is where [fine tuning](https://beta.openai.com/docs/guides/fine-tuning) can come in. A dataset can be prepared with a prompt, a single question and single desired response. Once fine tuned, we should be able to omit the examples.

> To my understanding, this "only" converts the properties.

So in this case yes, it only converts properties, because of it's lack of context into Rust it's difficult to create a more general model without the right dataset. 

>  I would choose different naming here and there, but I guess that is a matter of style and shouldn't be hard to change (Like selected instead of is_selected). But what about the actual implementation?

This would be a matter of preparing the right dataset and prompt, I feel this is one of the easier problems to solve. For example if we change the prompt to something like this, we may get better performance.

```
Convert the TypeScript Props to Yew Props written in the rust programming language. While converting prompts use the following guidelines.
1. Any react component types such as React.ReactNode or ReactNode can be typed as HTML in rust. 
2. TypeScript nullable types are typed as Optional in Rust. 
3. Visibility in Rust props are all public.  
4. Variable names should be converted from CamelCase to snake_case
5. Rust variable naming convention for boolean's is not prefixed with is or is_. 
```

This is where experimentation is valuable to perform A/B tests between different prompts.

> But what about the actual implementation? All those properties need to be applied to the virtual DOM model somehow. And that needs to fit into the actual logic of components. Could GPT-3 handle that too?

The way I think I would approach this, would be to parse the JS AST and have specific prompts for handling different parts of the component. I suspect anything around the state, props, JSX, or lifecycle shouldn't be too difficult. The actual logic of the component will be a challenge. 

OpenAI does have an [insert](https://beta.openai.com/docs/guides/code/inserting-code) and [edit](https://beta.openai.com/docs/guides/code/editing-code) api. Based on my knowledge my intuition tells me that if we can complete a first pass to create the skeleton of the component, there will be enough context for GPT to produce better results. Though it's not clear at this time how well it understands the virtual dom.

> In some cases I explicitly choose a different path, which felt more Rust native (like using enums for conflicting settings).

Good callout, I think this is easily doable with the right prompt and dataset. I think some of this might become more clear when we start talking about what a tool or product might look like.

#### Tooling

I see a short and long term solution. To start a simple CLI that takes a string argument or file path and it will contain a predefined prompt for the different subcommands, i.e props, state, any JSX etc. 

> What would the workflow look like?

Expanding on the CLI a simple web frontend can be built around this. The main body containing a left and right text area and a toolbar somewhere. A user would paste in a react component on the left, click one of the buttons in the toolbar and content in the right text area will be updated with the output. 

> Just an "initial" conversion and then manual tracking of changes? Or repeating the generating with every release?

Yeah, this will be a lot of manual tracking of changes though exploration. Over time the model can be fine tuned. We should probably define some goals around precision and manual work will decrease over time.

> What would a developer on this project require to re-generate the code by himself?

It will likely be a manual effort for the short term starting with the CLI and moving to the web UI. Once we have some goals defined it would make sense to look into CI and have it just address the upstream diff leaving a PR for review. (It's possible it can address PR comments in far future)

#### Other Questions

I still have a lot of open questions myself and they lead to many more. GPT is a general purpose transformer. It doesn't just write code, it can answer questions. 

Will fine tuning with Q&A from the Rust / Yew docs have an impact coding performance?

Will fine tuning with Patternfly Docs or code impact model performance?

Will fine tuning on existing Rust projects impact performance?

How does formatting the question impact performance?

Is this general enough to apply to any React Project?

Once we have it fine tuned and dialed into the syntax we like can it do it all in a single pass?

Can we provide a feedback loop with the Rust analyzer? Will it be able to fix any compiler errors it creates? Will this improve overall Rust performace?

Will providing additional documentation around the React virtual Dom improve performance and understand with Yew apps?

**Comment by ctron** [comment]
Yea, that all sounds pretty reasonable. And super interesting. I like the idea of exploring, and also having some manual process, with the user double checking the result. Also using the Rust analyzer (maybe clippy too) and PRs in a feedback loop sounds like the right thing.

I am not sure I can help with that tough. But I am definitely trying to support you.
